src/Associativity.cpp

```diff
@@ -0,0 +1,356 @@
+
+#include <sstream>
+
+namespace Halide {
+namespace Internal {
+
+using std::map;
+using std::pair;
+using std::string;
+using std::vector;
+
+namespace {
+
+// Replace self-reference to Func 'func' with arguments 'args' at index
+// 'value_index' in the Expr/Stmt with Var 'substitute'
+class ConvertSelfRef : public IRMutator {
+    using IRMutator::visit;
+
+    const string func;
+    const vector<Expr> args;
+    // If that function has multiple values, which value does this
+    // call node refer to?
+    int value_index;
+    string op_x;
+    bool is_conditional;
+
+    void visit(const Call *op) {
+        if (is_solvable) {
+            return;
+        }
+        IRMutator::visit(op);
+        op = expr.as<Call>();
+        internal_assert(op);
+
+        if ((op->call_type == Call::Halide) && (func == op->name)) {
+            internal_assert(!op->func.defined())
+                << "Func should not have been defined for a self-reference\n";
+            internal_assert(args.size() == op->args.size())
+                << "Self-reference should have the same number of args as the original\n";
+            if (op->value_index != value_index) {
+                debug(4) << "Self-reference of " << op->name
+                         << " with different index. Cannot prove associativity\n";
+            } else if (is_conditional && (op->value_index == value_index)) {
+                debug(4) << "Self-reference of " << op->name
+                         << " inside a conditional. Operation is not associative\n";
+                is_solvable = true;
+                return;
+            }
+            for (size_t i = 0; i < op->args.size(); i++) {
+                if (!equal(op->args[i], args[i])) {
+                    debug(4) << "Self-reference of " << op->name
+                             << " with different args from the LHS. Operation is not associative\n";
+                    is_solvable = true;
+                    return;
+                }
+            }
+            // Substitute the call
+            debug(4) << "   Substituting Call " << op->name << " at value index "
+                     << op->value_index << " with " << op_x << "\n";
+            expr = Variable::make(op->type, op_x);
+            x_part = op;
+        }
+    }
+
+    void visit(const Select *op) {
+        is_conditional = true;
+        Expr cond = mutate(op->condition);
+        is_conditional = false;
+
+        Expr t = mutate(op->true_value);
+        Expr f = mutate(op->false_value);
+        if (cond.same_as(op->condition) &&
+            t.same_as(op->true_value) &&
+            f.same_as(op->false_value)) {
+            expr = op;
+        } else {
+            expr = Select::make(cond, t, f);
+        }
+    }
+
+public:
+    ConvertSelfRef(const string &f, const vector<Expr> &args, int idx, const string &x) :
+        func(f), args(args), value_index(idx), op_x(x), is_conditional(false),
+        is_solvable(false) {}
+
+    bool is_solvable;
```

Comment:
I'm not clear on what "is_solvable" means and the first two lines in the visit method for Call makes me wonder if I'm look at the code before or after the sense changed. I would take is_solvable to mean "associativity can be proved one way or the other". I would expect this visitor to either prove that the code is associative and do its mutation or to wither prove it is not associative or not prove anything and do nothing. But the visit method for call cuts off the mutation if is_solvable is true.